### PR TITLE
Feature/date query

### PR DESCRIPTION
**Pull request by joeyblake**
Lets review what I have so far. I noticed some strangeness with multiple range filters when I was writing the tests so thats something that needs to be looked at for sure.

### CODE REVIEW

File: classes/class-ep-api.php [hunk 1]
```diff
@@ -511,6 +511,51 @@ public function put_mapping() {
...
+								"year" => array(
+									"type" => "integer",
+								),
+								"month" => array(
+									"type" => "integer",
+								),
+								"monthnum" => array(
+									"type" => "integer",
+								),
+								"m" => array(
```
Comment (tlovett1) [inline classes/class-ep-api.php]: What is the difference between `m` and `monthnum`?

File: classes/class-ep-api.php [hunk 2]
```diff
@@ -612,10 +658,29 @@ public function prepare_post( $post_id ) {
 		 * This filter is named poorly but has to stay to keep backwards compat
 		 */
 		$post_args = apply_filters( 'ep_post_sync_args', $post_args, $post_id );
-
 		return $post_args;
 	}
 
+	private function prepare_date_terms( $post_date_gmt ) {
```
Comment (tlovett1) [inline classes/class-ep-api.php]: Docblock please :)

File: classes/class-ep-api.php [hunk 3]
```diff
@@ -612,10 +658,29 @@ public function prepare_post( $post_id ) {
...
+		$timestamp = strtotime( $post_date_gmt );
+		$date_terms = array(
+			'year' => (int) date( "Y", $timestamp),
+			'month' => (int) date( "m", $timestamp),
+			'monthnum' => (int) date( "m", $timestamp),
+			'w' => (int) date( "W", $timestamp),
```
Comment (tlovett1) [inline classes/class-ep-api.php]: Why store this twice?

File: classes/class-ep-api.php [hunk 4]
```diff
@@ -835,6 +900,33 @@ public function format_args( $args ) {
...
+		if ( $date_filter = EP_WP_Date_Query::simple_es_date_filter( $args ) ) {
+			$filter['and'][] = $date_filter;
+			$use_filters = true;
+		}
+
+		/**
+		 * 'date_query' arg support.
+		 * 
+		 */
+		if ( ! empty( $args['date_query'] ) ) {
```
Comment (tlovett1) [inline classes/class-ep-api.php]: Kind of a weird scenario but if someone had the `date_query` param as well as the simple date stuff, what would happen?

File: classes/class-ep-wp-date-query.php [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+
+class EP_WP_Date_Query extends WP_Date_Query {
+
+	public function get_es_filter() {
+		$filter = $this->get_es_filter_for_clauses();
+
+		return $filter;
+	}
+
+	protected function get_es_filter_for_clauses() {
```
Comment (tlovett1) [inline classes/class-ep-wp-date-query.php]: I'm confused on this. Why are we returning a piece of a SQL query?

File: classes/class-ep-wp-date-query.php [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+					$filter_count = count( $clause_filter );
+					if ( ! $filter_count ) {
+						$filter_chunks['filters'][] = '';
+					} else {
+						$filter_chunks['filters'][] = $clause_filter;
+					}
+
+					// This is a subquery, so we recurse.
+				} else {
+					$clause_filter              = $this->get_es_filter_for_query( $clause, $depth + 1 );
```
Comment (tlovett1) [inline classes/class-ep-wp-date-query.php]: We definitely need a unit test for this one.

File: classes/class-ep-wp-date-query.php [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+					$filter_chunks['filters'][] = $clause_filter;
+				}
+			}
+		}
+
+		if ( empty( $relation ) ) {
```
Comment (tlovett1) [inline classes/class-ep-wp-date-query.php]: I assume it's a todo to implement other relations?

File: classes/class-ep-wp-date-query.php [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+
+		return $filter_array;
+	}
+
+	protected function get_es_filter_for_clause( $query, $parent_query ) {
+
+		// The sub-parts of a $where part.
+		$filter_parts = array();
+
+		$column = ( ! empty( $query['column'] ) ) ? esc_sql( $query['column'] ) : $this->column;
```
Comment (tlovett1) [inline classes/class-ep-wp-date-query.php]: Don't understand this? Why `esc_sql`?

File: classes/class-ep-api.php (hunk 1 above)
Comment (joeyblake) [inline classes/class-ep-api.php]: m is yearmonth YYYYMM, monthnum is the month number, month and monthnum are the same though

File: classes/class-ep-api.php (hunk 3 above)
Comment (joeyblake) [inline classes/class-ep-api.php]: We don't need to, but we'll need a translation so query params get mapped to the correct place. I'll add that.

**Comment by joeyblake** [comment]
@tlovett1 updated. The only question left was what happens if you mix a simple query with a date query. The answer is they are just combined in the filter, so the simple term gets lumped in with date query filter.

**Comment by tlovett1** [comment]
I'm going to merge into develop to do testing a cleanup.

**Pull request merged**